### PR TITLE
[ONNX] Add diagonal symbolic

### PR DESCRIPTION
**Pull request by shubhambhokare1**
Adds support for diagonal symbolic

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -5234,6 +5234,17 @@ def forward(self, x):
         self.run_test(TensorFactory(), x, test_with_inputs=[another_x],
                       input_names=["input_1"], dynamic_axes={"input_1": [0, 1, 2]})
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_diagonal(self):
+        class TensorFactory(torch.nn.Module):
+            def forward(self, x):
+                return torch.diagonal()
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5241]: The documentation says torch.diagonal requires an argument:
https://pytorch.org/docs/master/generated/torch.diagonal.html

Did you forget to pass in x as the first arg?

Does this test pass locally for you?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -5234,6 +5234,17 @@ def forward(self, x):
         self.run_test(TensorFactory(), x, test_with_inputs=[another_x],
                       input_names=["input_1"], dynamic_axes={"input_1": [0, 1, 2]})
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_diagonal(self):
+        class TensorFactory(torch.nn.Module):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5239]: I'd rename to `Diagonal` for clarity.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -5234,6 +5234,17 @@ def forward(self, x):
...
+            def forward(self, x):
+                return torch.diagonal()
+
+        x = torch.randn(2, 4, 5, 2)
+        another_x = torch.randn(5, 6, 7, 8)
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5244]: Please add a comment explaining why these other test inputs are useful.

File: torch/onnx/symbolic_opset13.py [hunk 4]
```diff
@@ -4,9 +4,10 @@
 # This file exports ONNX ops for opset 13
 import torch
 import torch.onnx.symbolic_helper as sym_help
-from torch.onnx.symbolic_helper import parse_args, _unimplemented
-from torch.onnx.symbolic_opset9 import overload_by_arg_count, _maybe_cast_reduce_op_input, nonzero, expand
-from torch.onnx.symbolic_opset11 import unsqueeze
+from torch.onnx.symbolic_helper import _squeeze_helper, parse_args, _unimplemented
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:7]: please address flake8 issues

Comment (garymm) [reply torch/onnx/symbolic_opset13.py:7]: And you may wish to use this to run flake8 locally when you commit:
https://github.com/pytorch/pytorch/blob/master/tools/git-pre-commit

You can add it to your local git hooks:
`ln -s ../../tools/git-pre-commit .git/hooks/pre-commit`

File: torch/onnx/symbolic_opset13.py [hunk 5]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+
+
+@parse_args("v", "i", "i", "i")
+def diagonal(g, self, offset, dim1, dim2):
+    dim1_size = size(g, self, dim=g.op("Constant", value_t=torch.LongTensor([dim1])))
+    dim2_size = size(g, self, dim=g.op("Constant", value_t=torch.LongTensor([dim2])))
+
+    # Create appropriate mask
+    mask_shape = [dim1_size, dim2_size]
+    mask_shape = g.op("Concat", *mask_shape, axis_i=0)
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:325]: Re-assigning a variable to one of a totally different type seems to me to harm readability.
Can we rewrite to:
`mask_shape = g.op("Concat", dim1_size, dim2_size, axis_i=0)`
?

File: torch/onnx/symbolic_opset13.py [hunk 6]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+
+    # dim1 and dim2 appended as a dimension at the end of the shape
+    rank = sym_help._get_tensor_rank(self)
+    if rank is not None:
+        axes = list(range(rank))
+        axes.remove(dim1)
+        axes.remove(dim2)
+        self = g.op("Transpose", self, perm_i=axes + [dim1, dim2])
+    else:
+        return _unimplemented("diagonal", "negative axis with unknown input rank")
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:337]: What does "negative axis" refer to?

File: torch/onnx/symbolic_opset13.py [hunk 7]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+
+    # Update mask
+    axes = list(range(rank))
+    mask_shape = [size(g, self,
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:341]: Same point on re-using mask_shape for variables of different types

File: torch/onnx/symbolic_opset13.py [hunk 8]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+    rank = sym_help._get_tensor_rank(self)
+    if rank is not None:
+        axes = list(range(rank))
+        axes.remove(dim1)
+        axes.remove(dim2)
+        self = g.op("Transpose", self, perm_i=axes + [dim1, dim2])
+    else:
+        return _unimplemented("diagonal", "negative axis with unknown input rank")
+
+    # Update mask
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:339]: please expand this comment to explain a bit more

File: torch/onnx/symbolic_opset13.py [hunk 9]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+    axes = list(range(rank))
+    mask_shape = [size(g, self, 
+                       dim=g.op("Constant", value_t=torch.LongTensor([axis]))) for axis in axes[:-2]]
+    mask_shape.append(g.op("Constant", value_t=torch.LongTensor([-1])))
+    mask_shape.append(g.op("Constant", value_t=torch.LongTensor([-1])))
+    mask_shape = g.op("Concat", *mask_shape, axis_i=0)
+    mask = expand(g, mask, mask_shape, None)
+    
+    # Multiply input and mask
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:348]: Please expand the comment to explain why the multiplication is happening. Just stating that there's a multiplication is not useful.

File: torch/onnx/symbolic_opset13.py [hunk 10]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+    result = g.op("Mul", self, mask)
+    result = sym_help._reducesum_helper(g, result, axes_i=[-1], keepdims_i=0)
+
+    # Calculate gather indices based on offset and dims
+    _offset = g.op("Constant", value_t=torch.LongTensor([offset]))
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:353]: Please name this more descriptively. Maybe `offset_op` or `offset_contant`

File: torch/onnx/symbolic_opset13.py [hunk 11]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+    if offset >= 0:
+        diag_size = g.op("Max", g.op("Min", dim1_size, g.op("Sub", dim2_size, _offset)),
+                         g.op("Constant", value_t=torch.LongTensor([0])))
+    else:
+        diag_size = g.op("Max", g.op("Min", g.op("Add", dim1_size, _offset), dim2_size),
+                         g.op("Constant", value_t=torch.LongTensor([0])))
+
+    gather_indices = ones(g, g.op("Concat", *[diag_size], axis_i=0), 4, None, None)
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:361]: is there a difference between
`g.op("Concat", *[diag_size], axis_i=0`
and
`g.op("Concat", diag_size, axis_i=0`

?

File: torch/onnx/symbolic_opset13.py [hunk 12]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+    gather_indices = g.op("CumSum", gather_indices, g.op("Constant", value_t=torch.LongTensor([0])))
+    gather_indices = g.op("Add", gather_indices, g.op("Constant", value_t=torch.LongTensor([abs(offset) - 1])))
+
+    gather_shape = [size(g, result,
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:365]: Same point on re-using a local variable to mean different things

File: torch/onnx/symbolic_opset13.py [hunk 13]
```diff
@@ -312,3 +313,62 @@ def repeat_interleave(g, self, repeats, dim=None, output_size=None):
...
+
+    gather_shape = [size(g, result, 
+                         dim=g.op("Constant", value_t=torch.LongTensor([axis]))) for axis in axes[:-2]]
+    gather_shape.append(g.op("Constant", value_t=torch.LongTensor([-1])))
+    gather_shape = g.op("Concat", *gather_shape, axis_i=0)
+
+    gather_indices = expand(g, gather_indices, gather_shape, None)
+    gather_indices = sym_help._unsqueeze_helper(g, gather_indices, [rank - 1])
+    
+    # Perform gather operation
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:373]: This comment is not really helpful.
Usually comments that just states *what* the code does are not helpful. The code should be readable enough that that's obvious from reading it.
Explaining *why* the code is doing something is much more often useful.

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks for the change!

File: torch/onnx/symbolic_opset13.py (hunk 11 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset13.py:361]: No difference, updated.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (shubhambhokare1) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:5241]: Added x as an argument

**Comment by garymm** [comment]
@shubhambhokare1 I haven't seen any changes to this PR. Did you forget to push?

File: torch/onnx/symbolic_opset13.py (hunk 9 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset13.py:348]: Added description

**Comment by shubhambhokare1** [comment]
@garymm Pushed latest changes.

File: torch/onnx/symbolic_opset13.py [hunk 14]
```diff
@@ -334,41 +333,62 @@ def diagonal(g, self, offset, dim1, dim2):
...
+    # There might be cases where offset value is greater than number of rows/columns
+    # and might cause the diagonal to overrun. In these cases, we return a tensor which
+    # consists of a 0-dim value
+    overrun_cond = g.op("Not", g.op("Equal", diag_size, g.op("Constant", value_t=torch.tensor(0, dtype=torch.int64))))
+    # create the "If" node and add the "then" and "else" blocks to it.
+    if_node_outputs = g.op("If", overrun_cond)
+    if_node = if_node_outputs.node()
+
+    if_block = _add_block(if_node)
+    gather_indices_ = if_block.op("Add", gather_indices, select_window)
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:386]: It's not clear from the variable name how `gather_indices_` and `gather_indices` differ. Please use a more descriptive name.

File: torch/onnx/symbolic_opset13.py [hunk 15]
```diff
@@ -334,41 +333,62 @@ def diagonal(g, self, offset, dim1, dim2):
...
-    # Perform gather operation
-    return g.op("GatherND", result, gather_indices, batch_dims_i=rank - 2)
+    gather_indices = zeros(g, gather_shape, 4, None, None)
+
+    # There might be cases where offset value is greater than number of rows/columns
+    # and might cause the diagonal to overrun. In these cases, we return a tensor which
+    # consists of a 0-dim value
+    overrun_cond = g.op("Not", g.op("Equal", diag_size, g.op("Constant", value_t=torch.tensor(0, dtype=torch.int64))))
+    # create the "If" node and add the "then" and "else" blocks to it.
+    if_node_outputs = g.op("If", overrun_cond)
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:382]: I suggest naming this `if_op` instead.

File: torch/onnx/symbolic_opset13.py [hunk 16]
```diff
@@ -334,41 +333,62 @@ def diagonal(g, self, offset, dim1, dim2):
...
     gather_shape = g.op("Concat", *gather_shape, axis_i=0)
-
-    gather_indices = expand(g, gather_indices, gather_shape, None)
-    gather_indices = sym_help._unsqueeze_helper(g, gather_indices, [rank - 1])
-    
-    # Perform gather operation
-    return g.op("GatherND", result, gather_indices, batch_dims_i=rank - 2)
+    gather_indices = zeros(g, gather_shape, 4, None, None)
+
+    # There might be cases where offset value is greater than number of rows/columns
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:377]: It's not obvious to me how the `diag_size != 0` relates to `offset value is greater than the number of rows/columns`. Could you change the comment or the condition to make it clearer?

Also minor grammatical suggestion for making the comment more concise:
Rewrite `There might be cases where $COND. In these cases, $CONSEQUENCE` to
`In cases where $COND, $CONSEQUENCE`.

**Review by garymm — CHANGES_REQUESTED** [review]

File: torch/onnx/symbolic_opset13.py [hunk 17]
```diff
@@ -334,41 +333,62 @@ def diagonal(g, self, offset, dim1, dim2):
...
+    # and might cause the diagonal to overrun. In these cases, we return a tensor which
+    # consists of a 0-dim value
```
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset13.py:379]: Added more details

File: torch/onnx/symbolic_opset13.py [hunk 18]
```diff
@@ -334,41 +333,62 @@ def diagonal(g, self, offset, dim1, dim2):
...
+    diag_size = g.op("Concat", diag_size, axis_i=0)
+
+    # Calculate which diagonal values to select
+    # For example, in cases with offsets:
+    # [[0, 1.1, 0]
+    #  [0, 0, 2.2]]
+    # we need to select the last two columns, so we create a tensor
+    # with all columns that are to be selected
+    # So in this example, it is [1, 2]
+    select_window_ones_fill = ones(g, diag_size, 4, None, None)
```
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset13.py:367]: I think this comes from the symbolic_helper.py
```scalar_type_to_pytorch_type = [
    torch.uint8,        # 0
    torch.int8,         # 1
    torch.short,        # 2
    torch.int,          # 3
    torch.int64,        # 4
    torch.half,         # 5
    torch.float,        # 6
    torch.double,       # 7
    torch.complex32,    # 8
    torch.complex64,    # 9
    torch.complex128,   # 10
    torch.bool,         # 11
]```
in zeros/ones function, we call scalar_type_to_pytorch_type[dtype], in our case dtype is a number

There is a possibility in future PRs to replace this mechanism

[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_opset13.py:367]: Can you please file a work item to track that improvement?

File: torch/onnx/symbolic_opset13.py [hunk 19]
```diff
@@ -375,20 +375,27 @@ def diagonal(g, self, offset, dim1, dim2):
...
+    # and might cause the diagonal to overrun and as a result of this, diag_size would be zero.
+    # For example, if
+    #       offset = 9, dim1_size = 2 (columns), dim2_size = 4 (rows)
+    #       diag_size = max(min(2, (4-9)), 0) = 0, based on calculation above
+    # Cases with diagonal overrun always result in diag_size = max(0, -ve value) = 0
+    # In cases without diagonal overrun, we select the appropriate rows/columns along which we
+    # are calculating diagonal values. In cases with diagonal overrun, we return a tensor which has
+    # the dimension of the row/column where overrun occurred as 0-dim, as we are essentially
+    # returning an empty tensor
+
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:387]: please remove this empty line

**Review by garymm — APPROVED** [review]
Thanks!
